Mathlib/Algebra/Algebra/Subalgebra/Basic.lean

```diff
@@ -938,6 +938,18 @@ variable (f : A →ₐ[R] B)
 theorem range_comp_val : (f.comp S.val).range = S.map f := by
   rw [AlgHom.range_comp, range_val]
 
+/-- An `AlgHom` between two rings restricts to an `AlgHom` from any subalgebra of the
+domain onto the image of that subalgebra. -/
+def _root_.AlgHom.subalgebraMap : S →ₐ[R] S.map f :=
+  (equivOfEq _ _ (range_comp_val S f)).toAlgHom.comp (f.comp S.val).rangeRestrict
+
+variable {S} in
+@[simp]
+theorem _root_.AlgHom.subalgebraMap_coe_apply (x : S) : f.subalgebraMap S x = f x := rfl
+
+theorem _root_.AlgHom.subalgebraMap_surjective : Function.Surjective (f.subalgebraMap S) := by
+  simpa [AlgHom.subalgebraMap] using (f.comp S.val).rangeRestrict_surjective
```

Comment:
This works with your original def as well as my suggested one. Lean can't infer `S.toAddSubmonoid` so you have to specify it.
```suggestion
theorem _root_.AlgHom.subalgebraMap_surjective : Function.Surjective (f.subalgebraMap S) :=
  f.toAddMonoidHom.addSubmonoidMap_surjective S.toAddSubmonoid
```